brushfire-core/src/main/scala/com/stripe/brushfire/local/Trainer.scala

```diff
@@ -3,81 +3,176 @@ package com.stripe.brushfire.local
 import com.stripe.brushfire._
 import com.twitter.algebird._
 
-case class Trainer[K: Ordering, V, T: Monoid](
-    trainingData: Iterable[Instance[K, V, T]],
-    sampler: Sampler[K],
-    trees: List[Tree[K, V, T]]) {
 
-  private def updateTrees(fn: (Tree[K, V, T], Int, Map[LeafNode[K, V, T, Unit], Iterable[Instance[K, V, T]]]) => Tree[K, V, T]): Trainer[K, V, T] = {
-    val newTrees = trees.zipWithIndex.par.map {
-      case (tree, index) =>
-        val byLeaf =
-          trainingData.flatMap { instance =>
-            val repeats = sampler.timesInTrainingSet(instance.id, instance.timestamp, index)
-            if (repeats > 0) {
-              tree.leafFor(instance.features).map { leaf =>
-                1.to(repeats).toList.map { i => (instance, leaf) }
-              }.getOrElse(Nil)
-            } else {
-              Nil
-            }
-          }.groupBy { _._2 }
-            .mapValues { _.map { _._1 } }
-        fn(tree, index, byLeaf)
-    }
-    copy(trees = newTrees.toList)
+trait TrainingStep[K,V,T] {
+
+  type K1
+  type V1
+  type V2
+
+  def prepare(trees: Map[Int, Tree[K,V,T]], instance: Instance[K,V,T]): Seq[((Int,Int,K1), V1)]
+  def lift(tree: Tree[K,V,T], leafIndex: Int, key: K1, v1: V1): Traversable[V2]
+  def update(tree: Tree[K,V,T], map: Map[Int,V2]): Tree[K,V,T]
```

Comment:
Hmmm. Well, two things:

- Absent the inverted tree representation, growing a tree is kinda expensive, and so it's nicer to do it all at once. (With the inverted tree representation it would be fine to do it one by one).
- For pruning, rather than expanding/annotating the leaves, you actually do need to do it all at once for the whole tree, so if we had a leaf by leaf update method here we couldn't implement prune.